### ISSUE TITLE
`Failed to init template entry.kte, no method named 'render' found in class gg.jte.generated.precompiled.JteentryGenerated` when argument is Kotlin value class

### ISSUE THREAD

**Issue by asm0dey** [casid/jte#163]
Hello,

I use jte 2.1.1 inside my ktor application with Gradle.
Gradle setup is following:

```kotlin
jte {
    generate()
    sourceDirectory.set(Paths.get("templates"))
    contentType.set(gg.jte.ContentType.Html)
    binaryStaticContent.set(true)
    trimControlStructures.set(true)
}
```

Most part of templates work fine, but for one of them, namely `entry.kte` seemingly there is no "render" method. At least during debug I can't find it too. Here's the `entry.kte` content:
```
@import io.github.asm0dey.plugins.*
@import java.time.*

@param content: String
@param summary: String
@param book: BookWithInfo
@param imageType: String
<?xml version="1.0" encoding="UTF-8"?>
<entry xmlns="http://www.w3.org/2005/Atom" xmlns:dcterms="http://purl.org/dc/terms/">
    <title>${book.book.name}</title>
    <id>/opds/book/${book.id}/info</id>
    @for(author in book.authors)
    <author>
        <name>${author.buildName()}</name>
        <uri>/opds/author/browse/${author.id}</uri>
    </author>
    @endfor
    <published>${formatDate(ZonedDateTime.now())}</published>
    <updated>${formatDate(book.book.added)}</updated>
    @if(book.book.lang != null)
    <dcterms:language>${book.book.lang}</dcterms:language>
    @endif
    @if(book.book.date != null)
    <dcterms:date>${book.book.date}</dcterms:date>
    @endif
    @for(genre in book.genres)
    <category term="${genre}" label="${genre}"/>
    @endfor
    @if(imageType != null)
    <link type="${imageType}" rel="http://opds-spec.org/image" href="/opds/image/${book.id}"/>
    @endif
    <link type="application/fb2+zip" rel="http://opds-spec.org/acquisition/open-access" href="/opds/book/${book.id}/download"/>
    @if(summary != null)
    <summary type="text">${summary}</summary>
    <content type="html">$unsafe{content}</content>
    @endif
</entry>
```

Looking at javap output I can see following:
```
❯ javap ./build/classes/kotlin/main/gg/jte/generated/precompiled/JteentryGenerated.class | grep render
  public static final void render-JMhnnco(gg.jte.html.HtmlTemplateOutput, gg.jte.html.HtmlInterceptor, java.lang.String, java.lang.String, org.jooq.Record4<java.lang.Long, java.util.List<? extends io.github.asm0dey.opdsko.jooq.tables.pojos.Book>, java.util.List<? extends io.github.asm0dey.opdsko.jooq.tables.pojos.Author>, java.util.List<? extends java.lang.String>>, java.lang.String);
  public static final void renderMap(gg.jte.html.HtmlTemplateOutput, gg.jte.html.HtmlInterceptor, java.util.Map<java.lang.String, ? extends java.lang.Object>);
```

_Further investigation:_

`BookWithInfo` is a `value` class:
```kotlin
@JvmInline
value class BookWithInfo(private val record: Record4<Long, List<BookPojo>, List<Author>, List<String>>) {
    val book get() = record.component2()[0]
    val authors get() = record.component3()!!
    val genres get() = record.component4().map { genreNames[it] ?: it }
    val id get() = record.get(BOOK.ID)!!
}
```
Maybe this is the reason why we can see a bridge-like `render-JMhnnco` method in the `JteentryGenerated.class`?

However, for other template, where `BookWithInfo` participates too, but in a `List`, not by itself, everything is being generated fine:

```
❯ javap ./build/classes/kotlin/main/gg/jte/generated/precompiled/JtebooksGenerated.class | grep render
  public static final void render(gg.jte.html.HtmlTemplateOutput, gg.jte.html.HtmlInterceptor, java.util.List<io.github.asm0dey.plugins.BookWithInfo>, java.util.Map<java.lang.Long, java.lang.String>, java.util.Map<java.lang.Long, java.lang.String>, java.lang.String, java.lang.String, java.lang.String, java.time.temporal.TemporalAccessor, java.util.List<io.github.asm0dey.model.Entry$Link>);
  public static final void renderMap(gg.jte.html.HtmlTemplateOutput, gg.jte.html.HtmlInterceptor, java.util.Map<java.lang.String, ? extends java.lang.Object>);
```

Maybe this is an error on Kotlin compiler's side, but it appears to me that template detection method should be just a bit more sophisticated and be aware that Kotlin value classes exist as well as such bridge methods.

**Comment by asm0dey** [comment]
When I change `value` class to a regular everything works fine

**Comment by asm0dey** [comment]
Definitely related Kotlin issue: https://youtrack.jetbrains.com/issue/KT-47338

**Comment by casid** [comment]
Welcome @asm0dey, and thanks for reporting.

Interesting, I didn't know that Kotlin is doing this. At first sight, this looks like a problem with the Kotlin compiler.

I'm wondering if `@JvmName("render")` added to the generated template code would help with the issue. Will try to reproduce this and give it a shot.

**Comment by casid** [comment]
Unfortunately disabling the mangling did not work as planned :-D (see https://github.com/casid/jte/commit/5a4d6f5ecde228ad4435a5493bf4d31bd1da1d43)

```
Failed to render helloValueClass.kte, type mismatch for parameter: Expected java.lang.String, got test.ValueClass
gg.jte.TemplateException: Failed to render helloValueClass.kte, type mismatch for parameter: Expected java.lang.String, got test.ValueClass
```

The method name is now unchanged, but now the inline type is required.

**Comment by asm0dey** [comment]
You're welcome, it was an interesting half an hour of debugging :)
Yes, I'm kinda surprised with this behavior too :)

вс, 26 июн. 2022 г., 20:48 Andreas Hager ***@***.***>:

**Comment by asm0dey** [comment]
And yes, it looks like it should be possible to work this around with Kotlin reflection. If an argument is a value type we should expect internal class there. But the next question is methods of this class. Don't have any idea how to handle it...

**Comment by casid** [comment]
So I gave this some more thoughts.

An option could be to wait for project Valhalla and hope for better interoperability of value classes between Java and Kotlin. This would mean that value classes are not supported in kte templates and the engine should provide the user with better error information than "method not found".

**Comment by asm0dey** [comment]
Yes it's definitely an option. Maybe you could detect this renamed method
and based on it's existence say that value classes are not supported. Also
maybe it makes sense to mention it in the documentation.

вс, 3 июл. 2022 г., 09:30 Andreas Hager ***@***.***>:

**Comment by casid** [comment]
While working on the error message I think I found a workaround for the issue. When using the renderMap() method and passing the value class in a map I got the test to pass.

https://github.com/casid/jte/commit/0c31c041c08d466281bcc611bd4f1212d0a34307#diff-3e18f24de83f4aaf2203f0a26a680c6119ba37d8a8ef2ac37d716ebb813f19b8

There's now also a helpful error message for when the value class is passed as single parameter and there's a type mismatch due to the Kotlin issue you've linked.

I'm going to write another test for value classes passed to nested templates, but since those template invocations are all done by the Kotlin compiler I'm quite sure this will work.

**Comment by asm0dey** [comment]
Impressive message!

**Comment by casid** [comment]
Thanks! Would be a shame if future users would need to spend time debugging :-)

I just added another test, where a template with a value class parameter calls another template also with a value class parameter. At least locally, this works!

I'll wait for the CI to pass and then merge this into main.

**Comment by casid** [comment]
Merged! 

Would you like to give it a try in your project? For that you'd need to clone the jte repo, do a `mvn clean install` on jte-parent and temporarily reference version `2.1.2-SNAPSHOT` in your project.

**Comment by asm0dey** [comment]
Seemingly it works! You can find the build here: https://jitpack.io/#casid/jte/fa73ea1a9d (I hate `mvn install` when I have other options :)

**Comment by asm0dey** [comment]
Generated bytecode includes 
```
  public static final void render-JMhnnco(gg.jte.html.HtmlTemplateOutput, gg.jte.html.HtmlInterceptor, java.lang.String, java.lang.String, org.jooq.Record4<java.lang.Long, java.util.List<? extends io.github.asm0dey.opdsko.jooq.tables.pojos.Book>, java.util.List<? extends io.github.asm0dey.opdsko.jooq.tables.pojos.Author>, java.util.List<? extends java.lang.String>>, java.lang.String);
```

**Comment by casid** [comment]
> Seemingly it works! You can find the build here: https://jitpack.io/#casid/jte/fa73ea1a9d (I hate `mvn install` when I have other options :)

Oh, that's really convenient, thanks for sharing this trick!

Thank you for testing & confirming the fix. I'm glad we got it working :-)

**Comment by asm0dey** [comment]
Me too! Thank you for help with investigation and effort to fix!

вс, 3 июл. 2022 г., 12:44 Andreas Hager ***@***.***>:

**Comment by casid** [comment]
I just released version [2.1.2](https://github.com/casid/jte/releases/tag/2.1.2) to maven central, including this fix :-)
